osu.Game.Rulesets.Osu.Tests/SpinFramesGenerator.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using osu.Game.Rulesets.Osu.Replays;
+using osu.Game.Rulesets.Replays;
+using osuTK;
+
+namespace osu.Game.Rulesets.Osu.Tests
+{
+    public class SpinFramesGenerator
```

Comment:
Because this is a class only used in tests, I would hope that structurally it's very easy to quickly understand.

Due to the self-nesting, right now I find things very hard to follow. Could it be restructured to not be nested?

ie.

```csharp

var generator = new SpinFramesGenerator(double startTime)
generator.Spin(from: 0, to: 60, duration: 1000);
var frames = generator.Build();
```

Or if you're wanting to keep the chaining,

```csharp
var generator = new SpinFramesGenerator(double startTime)
    .Spin(from: 0, to: 60, duration: 1000)
    .Spin(...)
    .Build();
```

should still be possible without the nesting by returning `this`.

Reply:
Yeah, definitely agree with this. The linked-list-y structure here isn't really doing much here except obfuscating intent. It shouldn't take very much rearranging at all to get rid of it in favour of collecting to like an internal list or something.

Reply:
Done.